### PR TITLE
Fix tests that fail when find closes and reopens

### PR DESCRIPTION
**Pull request by cwrw**
## Context

Tests are failing when cycle dates change

## Changes proposed in this pull request

make sure tests don't fail when cycle switches

## Guidance to review

I ran now, when find closes, find opens and apply opens and all passes

## Things to check

- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: spec/components/candidate_interface/grouped_provider_courses_component_spec.rb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe CandidateInterface::GroupedProviderCoursesComponent do
+RSpec.describe CandidateInterface::GroupedProviderCoursesComponent, mid_cycle: true do
```
Comment (duncanjbrown) [inline spec/components/candidate_interface/grouped_provider_courses_component_spec.rb:3]: The second spec (of two) in this file explicitly travels to the correct point in the cycle — probably clearer to travel to `CycleTimetable.find_opens` in the first one?

File: spec/models/apply_again_feature_metrics_spec.rb [hunk 2]
```diff
@@ -51,7 +51,7 @@ def reject(application_form)
       end
 
       it 'returns 1.0 when 100% of apply again applications are successful within given time range' do
-        @today = Time.zone.local(2020, 12, 31, 12)
+        @today = Time.zone.local((RecruitmentCycle.current_year - 1), 12, 31, 12)
```
Comment (duncanjbrown) [inline spec/models/apply_again_feature_metrics_spec.rb:54]:
```suggestion
        @today = Time.zone.local((RecruitmentCycle.previous_year), 12, 31, 12)
```

Here and below?

File: spec/services/carry_over_application_spec.rb [hunk 3]
```diff
@@ -161,12 +167,14 @@ def original_application_form
       before { FeatureFlag.activate(:restructured_immigration_status) }
 
       it 'sets the personal details section to incomplete if nationality is not UK/Irish for 2021-22 carry over' do
-        original_application_form.update!(
-          first_nationality: 'Indian',
-          personal_details_completed: true,
-          recruitment_cycle_year: 2021,
-        )
-        described_class.new(original_application_form).call
+        ApplicationForm.with_unsafe_application_choice_touches do
```
Comment (duncanjbrown) [inline spec/services/carry_over_application_spec.rb:170]: :+1:

File: spec/system/reject_by_default_spec.rb [hunk 4]
```diff
@@ -47,8 +47,8 @@ def and_an_application_is_ready_to_reject_by_default
   end
 
   def when_the_application_is_getting_close_to_the_reject_by_default_date
-    time_limit_in_buiness_days = TimeLimitCalculator.new(rule: :chase_provider_before_rbd, effective_date: Time.zone.now).call[:days]
-    Timecop.travel((time_limit_in_buiness_days.days - 1).before(@application_choice.reject_by_default_at)) do
+    time_limit_in_buiness_days = TimeLimitCalculator.new(rule: :chase_provider_before_rbd, effective_date: Time.zone.now).call[:time_in_future]
```
Comment (duncanjbrown) [inline spec/system/reject_by_default_spec.rb:50]: Typo! and could this var have a better name?

File: spec/workers/teacher_training_public_api/sync_all_providers_and_courses_worker_spec.rb [hunk 5]
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe TeacherTrainingPublicAPI::SyncAllProvidersAndCoursesWorker do
+RSpec.describe TeacherTrainingPublicAPI::SyncAllProvidersAndCoursesWorker, mid_cycle: true do
```
Comment (duncanjbrown) [inline spec/workers/teacher_training_public_api/sync_all_providers_and_courses_worker_spec.rb:3]: This feels a bit weird in this context?

Comment (duncanjbrown) [reply spec/workers/teacher_training_public_api/sync_all_providers_and_courses_worker_spec.rb:3]: oops wrong file!

File: spec/workers/start_of_cycle_notification_worker_spec.rb [hunk 6]
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe StartOfCycleNotificationWorker do
+RSpec.describe StartOfCycleNotificationWorker, mid_cycle: true do
```
Comment (duncanjbrown) [inline spec/workers/start_of_cycle_notification_worker_spec.rb:3]: Does this feel a bit weird in this context?

File: spec/workers/open_all_courses_on_apply_worker_spec.rb [hunk 7]
```diff
@@ -29,7 +29,7 @@
   it 'wont open courses if Apply is not in the new cycle' do
     course = create(:course, open_on_apply: false)
 
-    Timecop.freeze(CycleTimetable.apply_reopens - 1.day) do
+    Timecop.freeze(CycleTimetable.apply_reopens(2022) - 10.days) do
```
Comment (duncanjbrown) [inline spec/workers/open_all_courses_on_apply_worker_spec.rb:32]: Feels a bit arbitrary — why 10?

**Review by duncanjbrown — APPROVED** [review]
:clap: ⭐  **Brilliant PR** — thank you for doing this pre-emptively, it's going to make everyone's lives a lot easier tomorrow. Nits only inline, merge as you like.

Next let's do the same for Apply reopening??

**Comment by cwrw** [comment]
> Next let's do the same for Apply reopening??

I also ran it with apply reopening and seems to pass as well, so far

**Pull request merged**